### PR TITLE
Emg peak model PR2

### PR DESCRIPTION
**Pull request by pcolaianni**
This should be a final version of the class.

If it looks good to you, I'll remove the "geogebra" helper function in `_test.cpp` file.
Regarding the `std::cout`, I implemented a "debug level" class parameter (named `"print_debug"`) so the user can select how much info he desired to see. I can also get rid of all of this, but I'd like to do so in future, once you had a chance to test this class with some real data and see how the results look.

### CODE REVIEW

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h [hunk 1]
```diff
@@ -567,6 +630,67 @@ namespace OpenMS
...
+      parameters with gradient descent. The decision of which points to skip is based on the
+      derivatives between consecutive points.
+
+      It first selects all those points whose intensity is below a certain value (`intensity_threshold`).
+      Then, the derivatives of all the remaining points are computed. Based on the results,
+      the algorithm selects those points that present a high enough derivative.
+      Once a low value is found, the algorithm stops taking points from that side.
+      It then repeats the same procedure on the other side of the peak.
+      If points are actually skipped, these will be found in the middle part of the peak,
+      possibly being the saturated region.
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h]: Replace lines 645-646 with the following:  "The goal is to limit the inclusion of saturated or spurious points near the peak apex during training."

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h [hunk 2]
```diff
@@ -623,6 +747,362 @@ namespace OpenMS
...
+      double& param_update,
+      double& param,
+      const double current_E,
+      const double previous_E
+    ) const;
+
+    /**
+      @brief Compute the cost given by loss function E
+
+      Needed by the gradient descent algorithm. The mean squared error is used.
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h]: "The mean squared error is used as the loss function E"

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h [hunk 3]
```diff
@@ -623,6 +747,362 @@ namespace OpenMS
...
+
+    /// Maximum number of gradient descent iterations in `fitEMGPeakModel()`
+    UInt max_gd_iter_;
+
+    /**
+      Whether additional points should be added when fitting EMG peak model,
+      particularly useful with cutoff peaks
+    */
+    bool compute_additional_points_;
+  };
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h:1017]: Docstrings looks nice.

File: src/tests/class_tests/openms/source/PeakIntegrator_test.cpp [hunk 4]
```diff
@@ -135,6 +141,69 @@ constexpr const char* BASELINE_TYPE_BASETOBASE = "base_to_base";
...
+}
+for (Size i = 0; i < cutoff_pos_min.size(); ++i)
+{
+  cutoff_chrom_min.push_back(ChromatogramPeak(cutoff_pos_min[i], cutoff_int[i]));
+  cutoff_chrom_sec.push_back(ChromatogramPeak(cutoff_pos_sec[i], cutoff_int[i]));
+  cutoff_spec_min.push_back(Peak1D(cutoff_pos_min[i], cutoff_int[i]));
+  cutoff_spec_sec.push_back(Peak1D(cutoff_pos_sec[i], cutoff_int[i]));
+}
+
+
```
Comment (dmccloskey) [inline src/tests/class_tests/openms/source/PeakIntegrator_test.cpp:206]: Just to double check, you also include the "ideal" peak in your tests?

File: src/tests/class_tests/openms/source/PeakIntegrator_test.cpp [hunk 5]
```diff
@@ -155,6 +224,9 @@ START_SECTION(getParameters())
   Param params = ptr->getParameters();
   TEST_EQUAL(params.getValue("integration_type"), INTEGRATION_TYPE_INTENSITYSUM)
   TEST_EQUAL(params.getValue("baseline_type"), BASELINE_TYPE_BASETOBASE)
+  TEST_EQUAL(params.getValue("print_debug"), 0)
```
Comment (dmccloskey) [inline src/tests/class_tests/openms/source/PeakIntegrator_test.cpp:227]: I think the use of the parameter to specify the verbosity is fine for now.  Like you said, we can decide later to remove it if need be.

File: src/tests/class_tests/openms/source/PeakIntegrator_test.cpp [hunk 6]
```diff
@@ -848,6 +920,459 @@ START_SECTION(PeakShapeMetrics calculatePeakShapeMetrics(
...
+  TOLERANCE_RELATIVE(1.0 + 1e-2)
+  for (Size i = 0; i < out_min.size(); i += 9)
+  {
+    TEST_REAL_SIMILAR(out_min[i].getIntensity(), out_sec[i].getIntensity())
+  }
+  TOLERANCE_RELATIVE(1.0 + 1e-5)
+
+  fda_emg = &out_min.getFloatDataArrays()[0];
+  TEST_EQUAL(fda_emg->getName(), "emg_parameters")
+  TEST_REAL_SIMILAR((*fda_emg)[0], 15515900)
```
Comment (dmccloskey) [inline src/tests/class_tests/openms/source/PeakIntegrator_test.cpp:985]: Are the order of the parameters documented somewhere?  You mentioned that you store them in one of the .h docstrings, but I did not see mention of which index corresponds to which parameters.

File: src/tests/class_tests/openms/source/PeakIntegrator_test.cpp [hunk 7]
```diff
@@ -848,6 +920,459 @@ START_SECTION(PeakShapeMetrics calculatePeakShapeMetrics(
...
+
+  params.setValue("compute_additional_points", "true");
+  pi_f.peakIntegrator.setParameters(params);
+  pi_f.emg_vector(saturated_cutoff_pos_min, h, mu, sigma, tau, out_xs, out_ys);
+  TEST_EQUAL(out_xs.size(), 71) // more points than before
+  TEST_REAL_SIMILAR(out_xs.front(), 14.2717555076923) // peak was cutoff on the left side
+  TEST_REAL_SIMILAR(out_ys.front(), 108845.941990663)
+}
+END_SECTION
+
```
Comment (dmccloskey) [inline src/tests/class_tests/openms/source/PeakIntegrator_test.cpp:1375]: Nice work adding test coverage for all of the private/helper functions.  That will help greatly when future researchers look at this code, and for when/if we need to make changes/modifications.

**Review by dmccloskey — CHANGES_REQUESTED** [review]
Looks good.  Did you implement the check for the case of extending a peak side out too far as we spoke about?  I did not see any tests covering or checking for such a case.

File: src/tests/class_tests/openms/source/PeakIntegrator_test.cpp (hunk 4 above)
Comment (pcolaianni) [reply src/tests/class_tests/openms/source/PeakIntegrator_test.cpp:206]: Yes. In those cases, theidentifiers are: `chromatogram`, `spectrum`, `position` and `intensity`, depending on which data structure I'm testing (a `MSChromatogram`, `MSSpectrum`, and `double`s)

**Comment by pcolaianni** [comment]
@dmccloskey Updated the PR. How does it look?

File: src/tests/class_tests/openms/source/PeakIntegrator_test.cpp (hunk 6 above)
Comment (pcolaianni) [reply src/tests/class_tests/openms/source/PeakIntegrator_test.cpp:985]: Added some doc in `fitEMGPeakModel()` docstrings. How does it look?

**Comment by pcolaianni** [comment]
Oh, did not see the summary comment. About the peak extending limit, there is no explicit test. But it is kind of embedded in other tests because once I had implemented the fix, I had to update some tests on output sizes (then being smaller than before the fix).
example: the case of the cutoff peak. Before the fix, it would extend to 40 points. Now it's just 28. (line 1000 of `_test.cpp` file)

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h [hunk 8]
```diff
@@ -528,8 +528,10 @@ namespace OpenMS
       to produce a peak with similar intensities on boundaries' points.
 
       Metadata will be added to the output peak, containing the optimal parameters
-      for the EMG peak model. These information will be found in a `FloatDataArray`
-      with name "emg_parameters".
+      for the EMG peak model. This information will be found in a `FloatDataArray`
+      of name "emg_parameters", with the parameters being saved in the following
+      order (from index 0 to 3): amplitude `h`, mean `mu`, standard deviation `sigma`,
+      exponent relaxation time `tau`.
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h]: Great.

File: src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h [hunk 9]
```diff
@@ -642,8 +646,8 @@ namespace OpenMS
       the algorithm selects those points that present a high enough derivative.
       Once a low value is found, the algorithm stops taking points from that side.
       It then repeats the same procedure on the other side of the peak.
-      If points are actually skipped, these will be found in the middle part of the peak,
-      possibly being the saturated region.
+      The goal is to limit the inclusion of saturated or spurious points near the
+      peak apex during training.
```
Comment (dmccloskey) [inline src/openms/include/OpenMS/ANALYSIS/OPENSWATH/PeakIntegrator.h]: Nice.

**Review by dmccloskey — APPROVED** [review]
Looks good.  Let's see what the other developers at OpenMS think.

**Pull request closed without merging**